nvflare/app_common/tie/controller.py

```diff
@@ -0,0 +1,565 @@
+        # first trigger the abort_signal to tell all components (mainly the controller's control_flow and connector)
+        # that check this signal to abort.
+        if self.abort_signal:
+            self.abort_signal.trigger(value=True)
+
+        # if there is error, call system_panic to terminate the job with proper status.
+        # if no error, the job will end normally.
+        if error:
+            self.system_panic(reason=error, fl_ctx=fl_ctx)
+
+    def _is_stopped(self):
+        # check whether the abort signal is triggered
+        return self.abort_signal and self.abort_signal.triggered
+
+    def _update_client_status(self, fl_ctx: FLContext, op=None, client_done=False):
+        """Update the status of the requesting client.
+
+        Args:
+            fl_ctx: FL context
+            op: the XGB operation requested
+            client_done: whether the client is done
+
+        Returns: None
+
+        """
+        with self.status_lock:
+            peer_ctx = fl_ctx.get_peer_context()
+            if not peer_ctx:
+                self.log_error(fl_ctx, "missing peer_ctx from fl_ctx")
+                return
+            if not isinstance(peer_ctx, FLContext):
+                self.log_error(fl_ctx, f"expect peer_ctx to be FLContext but got {type(peer_ctx)}")
+                return
+            client_name = peer_ctx.get_identity_name()
+            if not client_name:
+                self.log_error(fl_ctx, "missing identity from peer_ctx")
+                return
+            status = self.client_statuses.get(client_name)
+            if not status:
+                self.log_error(fl_ctx, f"no status record for client {client_name}")
+            assert isinstance(status, _ClientStatus)
+            if op:
+                status.last_op = op
+            if client_done:
+                status.app_done = client_done
+            status.last_op_time = time.time()
+
+    def _process_client_done(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareable:
+        """Process the ClientDone report for a client
+
+        Args:
+            topic: topic of the message
+            request: request to be processed
+            fl_ctx: the FL context
+
+        Returns: reply to the client
+
+        """
+        self.log_debug(fl_ctx, f"_process_client_done {topic}")
+        exit_code = request.get(Constant.MSG_KEY_EXIT_CODE)
+
+        if exit_code == 0:
+            self.log_info(fl_ctx, f"app client is done with exit code {exit_code}")
+        elif exit_code == Constant.EXIT_CODE_CANT_START:
+            self.log_error(fl_ctx, f"app client failed to start (exit code {exit_code})")
+            self.system_panic("app client failed to start", fl_ctx)
+        else:
+            # Should we stop here?
+            # Problem is that even if the exit_code is not 0, we can't say the job failed.
+            self.log_warning(fl_ctx, f"app client is done with exit code {exit_code}")
```

Comment:
I think we should take a more conservative approach and stop when the exit code is not 0?

What are the situations where we want it to continue?

Reply:
This should be determined by the nature of the external system design. In case of XGB, we know it's MPI based and hence fragile. But Flower may not be that fragile. 

In general, Flare cannot decide whether the job should be aborted if one client app failed.